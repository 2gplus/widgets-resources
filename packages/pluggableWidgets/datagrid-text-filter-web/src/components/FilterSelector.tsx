import { createElement, ReactElement, RefObject, useEffect, useRef, useState } from "react";

const options = [
    { value: "contains", label: "Contains" },
    { value: "startsWith", label: "Starts with" },
    { value: "endsWith", label: "Ends with" },
    { value: "greater", label: "Greater" },
    { value: "greaterEqual", label: "Greater equal" },
    { value: "equal", label: "Equal" },
    { value: "notEqual", label: "Not equal" },
    { value: "smaller", label: "Smaller" },
    { value: "smallerEqual", label: "Smaller equal" }
];

export function FilterSelector({ onChange }: { onChange: (value: string) => void }): ReactElement {
    const [value, setValue] = useState(options[0]);
    const [show, setShow] = useState(false);
    const listRef = useRef<HTMLUListElement>(null);
    useOnClickOutside(listRef, () => setShow(false));
    return (
        <div className="filter-selector">
            <div className="filter-selector-content">
                <button
                    className={`btn btn-default filter-selector-button ${value.value}`}
                    onClick={() => {
                        setShow(show => !show);
                    }}
                >
                    &nbsp;
                </button>
                {show && (
                    <ul className="filter-selectors" ref={listRef}>
                        {options.map((option, index) => (
                            <li
                                key={index}
                                onClick={() => {
                                    setValue(option);
                                    onChange(option.value);
                                }}
                                className={value.value === option.value ? "filter-selected" : ""}
                            >
                                {option.label}
                            </li>
                        ))}
                    </ul>
                )}
            </div>
        </div>
    );
}

export function useOnClickOutside(ref: RefObject<HTMLUListElement>, handler: () => void): void {
    useEffect(() => {
        const listener = (event: MouseEvent & { target: Node | null }): void => {
            if (!ref.current || ref.current.contains(event.target)) {
                return;
            }
            handler();
        };
        document.addEventListener("mousedown", listener);
        document.addEventListener("touchstart", listener);
        return () => {
            document.removeEventListener("mousedown", listener);
            document.removeEventListener("touchstart", listener);
        };
    }, [ref, handler]);
}
