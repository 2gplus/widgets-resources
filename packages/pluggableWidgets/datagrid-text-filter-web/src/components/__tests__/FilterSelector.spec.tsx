import { shallow } from "enzyme";
import { createElement } from "react";
import { FilterSelector } from "../FilterSelector";

describe("Filter selector", () => {
    it("renders correctly", () => {
        const component = shallow(<FilterSelector onChange={jest.fn()} />);

        expect(component).toMatchSnapshot();
    });
});
