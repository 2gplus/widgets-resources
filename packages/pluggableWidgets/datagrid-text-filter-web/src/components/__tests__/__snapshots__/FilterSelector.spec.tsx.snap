// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter selector renders correctly 1`] = `
<div
  className="filter-selector"
>
  <div
    className="filter-selector-content"
  >
    <button
      className="btn btn-default filter-selector-button contains"
      onClick={[Function]}
    >
      Â 
    </button>
  </div>
</div>
`;
